### ISSUE TITLE
Crash using Target batching on iOS

### ISSUE THREAD

**Issue by carllindberg** [Adobe-Marketing-Cloud/mobile-services#302]
We are seeing some crashes in version 4.17.  I recently switched to sometimes use target batch calls (even if just a single mbox, it can essentially be a batch of one), but I think these calls are still going through older APIs, even though it is getting a crash in ADBTargetBatchApiHandler.

These are calls to just an mbox name, with just mboxHost and one other parameter, and normally just gets a short text string back as the content.  They appear to be failing randomly (probably in bad networks -- I do see error logs for other target mbox calls nearby, but those aren't crashing).  It's possible we have some simultaneous requests going out.

Is there any workaround we can use for this to avoid these crashes?  They seem to happen if I use the newer batch APIs or the old ones.  AT least the "\<mbox name\>" part of the crash dump seems to be for calls either way.

```
Name	NSInvalidArgumentException
Reason	*** -[__NSDictionaryM setObject:forKey:]: object cannot be nil (key: <mbox name>)

0	CoreFoundation 0x000000018daf0ec4 __exceptionPreprocess + 224
1	libobjc.A.dylib 0x000000018ccc1a40 objc_exception_throw + 52
2	CoreFoundation 0x000000018da68494 _CFThrowFormattedException + 108
3	CoreFoundation 0x000000018d9dba30 -[__NSDictionaryM setObject:forKey:] + 928
!	4	<AppName> 0x00000001054b8224 -[ADBTargetBatchApiHandler processMboxesRequestJsonResponse:] + 600
5	<AppName> 0x000000010549476c __44+[ADBTarget loadRequests:profileParameters:]_block_invoke.148 + 184
6	<AppName> 0x00000001054a87a4 __28-[ADBNetworkObject sendSync]_block_invoke + 108
7	CFNetwork 0x000000018e0cda00 __75-[__NSURLSessionLocal taskForClass:request:uploadFile:bodyData:completion:]_block_invoke + 28
8	CFNetwork 0x000000018e0e21a0 __49-[__NSCFLocalSessionTask _task_onqueue_didFinish]_block_invoke + 172
9	Foundation 0x000000018e5808bc __NSBLOCKOPERATION_IS_CALLING_OUT_TO_A_BLOCK__ + 12
10	Foundation 0x000000018e488ab8 -[NSBlockOperation main] + 68
11	Foundation 0x000000018e487f8c -[__NSOperationInternal _start:] + 736
12	Foundation 0x000000018e582790 __NSOQSchedule_f + 268
13	libdispatch.dylib 0x000000018d5296c8 _dispatch_call_block_and_release + 20
14	libdispatch.dylib 0x000000018d52a484 _dispatch_client_callout + 12
15	libdispatch.dylib 0x000000018d500e04 _dispatch_continuation_pop$VARIANT$armv81 + 400
16	libdispatch.dylib 0x000000018d5005a4 _dispatch_async_redirect_invoke + 776
17	libdispatch.dylib 0x000000018d50caec _dispatch_root_queue_drain + 340
18	libdispatch.dylib 0x000000018d50d34c _dispatch_worker_thread2 + 112
19	libsystem_pthread.dylib 0x000000018d70d17c _pthread_wqthread + 468
20	libsystem_pthread.dylib 0x000000018d70fcec start_wqthread + 0
```

**Comment by PravinPK** [comment]
Hi @carllindberg,

Thanks for reporting the issue. We have found and fixed the bug. The fix will be available in our next release, which we are targeting for January 2019. Will keep you posted.

Thanks,
Pravin
Mobile software engineer
Adobe Marketing Cloud

**Comment by carllindberg** [comment]
Great, thanks.  Will look for it.

Is there anything I can do in the meantime, in case the crash count is a bit too much for us?  If I swizzle that processMboxes method to wrap the call in a try/catch, would that prevent processing from finishing, or would it just result in a more graceful failure?  Or would batching the calls more help?

**Comment by shalehaha** [comment]
@carllindberg  This issue has been fixed on version 4.18.0. Please upgrade the SDK and let us know if you have further questions.

**Comment by carllindberg** [comment]
Will confirm this did solve the crashes.  Thanks.
